kimchi/src/circuits/polynomials/poseidon.rs

```diff
@@ -1,73 +1,84 @@
-/*****************************************************************************************************************
-
-This source file implements Posedon constraint polynomials.
-
-*****************************************************************************************************************/
+//! This source file implements Posedon constraint polynomials.
 
 use crate::circuits::expr::{Cache, Column, ConstantExpr, E};
 use crate::circuits::gate::{CurrOrNext, GateType};
 use crate::circuits::gates::poseidon::*;
 use ark_ff::{FftField, SquareRootField};
 use oracle::poseidon::{PlonkSpongeConstants15W, SpongeConstants};
+use std::ops::RangeInclusive;
 use CurrOrNext::*;
 
-/// An equation of the form `(curr | next)[i] = round(curr[j])`
-pub struct RoundEquation {
-    pub source: usize,
-    pub target: (CurrOrNext, usize),
+/// All the information needed to construct a round in the Poseidon custom gate.
+pub struct RoundSpec {
+    /// the columns that contain the input
+    pub input_cols: RangeInclusive<usize>,
+    /// the row that contain the output
+    pub output_row: CurrOrNext,
+    /// the columns that contain the output
+    pub output_cols: RangeInclusive<usize>,
 }
 
-pub const ROUND_EQUATIONS: [RoundEquation; ROUNDS_PER_ROW] = [
-    RoundEquation {
-        source: 0,
-        target: (Curr, 1),
+/// Specifies in which columns the input of a row is stored,
+/// as well as which row and columns the output the permutation is stored.
+/// A Poseidon gates performs 5 rounds, with the last 5 round outputing
+/// its result on the next row. Each round acts on a state of 3 columns.
+/// The layout is also shuffled, so that the 4th round of the gate is stored
+/// right after the first round. This is to let the permutation access it,
+/// in case the final output lands on this state, instead of the next row's.
+pub const ROUND_EQUATIONS: [RoundSpec; ROUNDS_PER_ROW] = [
+    RoundSpec {
+        input_cols: 0..=2,
```

Comment:
Do you think it could make sense to predefine some ranges to assign these - otherwise hardcoded - fields? I am referring to the pattern/cycle for input and output columns: 0-2 -> 6-8 -> 9-11 -> 12-14 -> 3-5 -> 0-2

Reply:
I'm not sure I understand the question exactly. We could pre-compute these ranges based on a `STATE_SIZE`, but we're then losing the clean layout written in this form. One idea here was to remove the clever calculation and explicitly write out the layout.